File: DecisionTree/dTree.cpp
#include "library/headers.h"
#include "library/DecisionTree.h"

void SolveTask( int numberOfArgument, char *arguments[ ] ) {

        if( numberOfArgument != 5 ) {
                cout << "Incorrect number of arguments, please recheck !" << endl;
                return;
        }

        string trainingFile = arguments[ 1 ];
        string testDataFile = arguments[ 2 ];
        string trainigType  = arguments[ 3 ];
        string pruningThr   = arguments[ 4 ];
        stringstream ss( pruningThr );
        int pruningThreshold; ss >> pruningThreshold;

        if( trainigType  != "optimized" && trainigType != "randomized" && trainigType != "forest3" && trainigType != "forest15" ) {
                cout << "Training Type is not valid ! " << endl;
                return ;
        }

        DecisionTree dTree;
        dTree.SetPruningThreshold( pruningThreshold );
        dTree.Train( trainingFile, trainigType );
        dTree.PrintForest( "Forest.txt" );
        dTree.TestDataPrediction( testDataFile, "testResult.txt" );
}

int main( int argc, char *argv[ ] ) {

        SolveTask( argc, argv );

        return 0;
}

File: 2. Machine Learning/Supervised Learning/DecisionTree/python-impl/dTree.py
import sys
from library.DecisionTree import DecisionTree

def solve_task(number_of_arguments, arguments):
    if number_of_arguments != 5:
        print("Incorrect number of arguments, please recheck!")
        return

    training_file = arguments[1]
    test_data_file = arguments[2]
    training_type = arguments[3]
    pruning_thr = arguments[4]

    try:
        pruning_threshold = int(pruning_thr)
    except ValueError:
        print("Pruning threshold must be an integer!")
        return

    valid_training_types = ["optimized", "randomized", "forest3", "forest15"]
    if training_type not in valid_training_types:
        print("Training Type is not valid!")
        return

    d_tree = DecisionTree()
    d_tree.set_pruning_threshold(pruning_threshold)
    d_tree.train(training_file, training_type)
    #d_tree.print_forest("Forest.txt")
    d_tree.test_data_prediction(test_data_file, "testResult.txt")

if __name__ == "__main__":
    solve_task(len(sys.argv), sys.argv)
